node/src/components/linear_chain.rs

```diff
@@ -98,29 +100,75 @@ impl<I: Display> Display for Event<I> {
                 height,
                 block.is_some()
             ),
+            Event::GetBlockForFinalitySignaturesResult(block_hash, maybe_block) => {
+                write!(
+                    f,
+                    "linear chain get-block-for-finality-signatures-result for {} found: {}",
+                    block_hash,
+                    maybe_block.is_some()
+                )
+            }
         }
     }
 }
 
 #[derive(DataSize, Debug)]
 pub(crate) struct LinearChain<I> {
-    /// A temporary workaround.
-    // TODO: Refactor to proper LRU cache.
-    linear_chain: Vec<Block>,
+    /// The most recently added block.
+    latest_block: Option<Block>,
+    /// Finality signatures to be inserted in a block once it is available.
+    pending_finality_signatures: HashMap<PublicKey, HashMap<BlockHash, FinalitySignature>>,
```

Comment:
As @afck and I have discussed, this could contain more information which would be useful for synchronization.

Similar to as @afck suggests [above](https://github.com/CasperLabs/casper-node/pull/649/files#r540909572), we could add `block_height` to `FinalitySignature`. Then, we could add a enum.

```rust
enum FaultyOrFetchedFinalitySignature {
  /// A pair of finality signatures where:
  /// - `bad_sig1.public_key == bad_sig2.public_key`
  /// - `bad_sig1.block_height == bad_sig2.block_height`
  /// - `bad_sig1.block_hash != bad_sig2.block_hash`
  Faulty{
    bad_sig1: FinalitySignature, 
    bad_sig2: FinalitySignature
  },
  Fetched(FinalitySignature)
}
```

then `HashMap<PublicKey, HashMap<BlockHash, FinalitySignature>>` would become 

`HashMap<PublicKey, HashMap<BlockHash, FaultyOrFetchedFinalitySignature>>`

is this right @afck ?

Reply:
Yes, that makes sense to me. (Doesn't have to be in this PR either, I hope? I already told bors to merge. :grimacing: )

Reply:
If that's for synchronization purposes then it should be inside of `LinearChainSync` component and not `LinearChain`.

Reply:
Good point. We'll need to find a way to avoid duplicating all that information.